openvino_xai/methods/white_box/torch.py

```diff
@@ -0,0 +1,257 @@
+        """Process numpy input, return numpy output."""
+        if not self._model_compiled:
+            raise RuntimeError("Model is not compiled. Call prepare_model() first.")
+
+        if preprocess:
+            x = self.preprocess_fn(x)
+        x = torch.from_numpy(x).float()
+
+        with torch.no_grad():
+            x = self._model_compiled(x)
+
+        output = {}
+        for name, data in x.items():
+            if not isinstance(data, torch.Tensor):
+                data = torch.tensor(data)
+            output[name] = data.numpy(force=True)
+        return output
+
+    def _feature_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> torch.Tensor:
+        self._feature_map = output
+        return output
+
+    def _output_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> Dict[str, torch.Tensor]:
+        return {
+            "prediction": output,
+            SALIENCY_MAP_OUTPUT_NAME: torch.empty_like(output),
+        }
+
+    def generate_saliency_map(self, data: np.ndarray) -> np.ndarray:
+        """Return saliency map."""
+        model_output = self.model_forward(data)
+        return model_output[SALIENCY_MAP_OUTPUT_NAME]
+
+    @staticmethod
+    def _normalize_map(saliency_map: torch.Tensor) -> torch.Tensor:
+        """Normalize saliency maps."""
+        max_values = saliency_map.max(dim=-1, keepdim=True).values
+        min_values = saliency_map.min(dim=-1, keepdim=True).values
+        saliency_map = 255 * (saliency_map - min_values) / (max_values - min_values + 1e-12)
+        return saliency_map.to(torch.uint8)
+
+
+class ActivationMap(TorchMethod):
+    """ActivationMap. Mean of the feature map along the channel dimension."""
+
+    def _output_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> Dict[str, torch.Tensor]:
+        feature_map = self._feature_map
+        batch_size, _, h, w = feature_map.shape
+        activation_map = torch.mean(feature_map, dim=1)
+        if self._embed_scaling:
+            activation_map = activation_map.reshape((batch_size, h * w))
+            activation_map = self._normalize_map(activation_map)
+            activation_map = activation_map.reshape((batch_size, h, w))
+        return {
+            "prediction": output,
+            SALIENCY_MAP_OUTPUT_NAME: activation_map,
+        }
+
+
+class ReciproCAM(TorchMethod):
+    """Implementation of Recipro-CAM for class-wise saliency map.
+
+    Recipro-CAM: gradient-free reciprocal class activation map (https://arxiv.org/pdf/2209.14074.pdf)
+
+    :param optimize_gap: Whether to optimize out Global Average Pooling operation
+    :type optimizae_gap: bool
+    """
+
+    def __init__(self, *args, optimize_gap: bool = False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._optimize_gap = optimize_gap
+
+    def _feature_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> torch.Tensor:
```

Comment:
Doesn't `input` have type `torch.Tensor`?

Reply:
Inputs could be, as far as I know, tuple of tensors, a tensor or anything else so that we specify the exact types.